hazelcast/src/main/java/com/hazelcast/cache/impl/CacheSplitBrainHandlerService.java

```diff
@@ -144,8 +155,79 @@ private boolean isLocalPartition(int partitionId) {
 
         @Override
         public void run() {
-            final Semaphore semaphore = new Semaphore(0);
+            int recordCount = 0;
+            for (Map.Entry<String, Map<Data, CacheRecord>> recordMapEntry : recordMap.entrySet()) {
+                String cacheName = recordMapEntry.getKey();
+                Map<Data, CacheRecord> records = recordMapEntry.getValue();
+                Object mergePolicy = getCacheMergePolicy(cacheName);
+                if (mergePolicy instanceof SplitBrainMergePolicy) {
+                    // we cannot merge into a 3.9 cluster, since not all members may understand the MergeOperationFactory
+                    if (nodeEngine.getClusterService().getClusterVersion().isLessThan(Versions.V3_10)) {
+                        logger.info("Cannot merge map '" + cacheName
+                                + "' with merge policy '" + mergePolicy.getClass().getName()
+                                + "' until cluster is running version " + Versions.V3_10);
+                        continue;
+                    }
+                    recordCount += handleMerge(cacheName, records, (SplitBrainMergePolicy) mergePolicy, DEFAULT_BATCH_SIZE);
+                } else {
+                    recordCount += handleMerge(cacheName, records, (CacheMergePolicy) mergePolicy);
+                }
+            }
+            recordMap.clear();
+
+            try {
+                semaphore.tryAcquire(recordCount, recordCount * TIMEOUT_FACTOR, TimeUnit.MILLISECONDS);
+            } catch (InterruptedException e) {
+                logger.finest("Interrupted while waiting merge operation...");
+            }
+        }
+
+        private int handleMerge(String name, Map<Data, CacheRecord> recordMap, SplitBrainMergePolicy mergePolicy, int batchSize) {
+            Map<Address, List<Integer>> memberPartitionsMap = partitionService.getMemberPartitionsMap();
+
+            MutableLong[] counterPerMember = new MutableLong[partitionCount];
```

Comment:
I copied that whole logic from my `map.putAll()` optimizations.

This is a shared counter for all partitions of a member. The same `MutableLong counter = new MutableLong();` is put into `counterPerMember` for all partitions, which belong to the member. This counter is then used to check against the `batchSize`, so we can send an aggregated operation to that member, once we have collected `batchSize` entries, regardless to which partition they belong.

I hope the name is okay, otherwise it has to be fixed in `MapProxySupport` and the map split-brain PR as well :)